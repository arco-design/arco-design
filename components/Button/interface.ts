import React, { CSSProperties, HTMLProps, ReactNode } from 'react';

export interface BaseButtonProps {
  style?: CSSProperties;
  className?: string | string[];
  children?: ReactNode;
  /**
   * @zh
   * 按钮主要分为六种按钮类型：主要按钮、次级按钮、虚框按钮、文字按钮、线性按钮，`default` 为次级按钮。
   * @en
   * A variety of button types are available: `primary`, `secondary`, `dashed`,
   * `text`, `linear` and `default` which is the secondary.
   * @defaultValue default
   */
  type?: 'default' | 'primary' | 'secondary' | 'dashed' | 'text' | 'outline';
  /**
   * @zh 按钮状态
   * @en Status of the button
   * @defaultValue default
   */
  status?: 'warning' | 'danger' | 'success' | 'default';
  /**
   * @zh 按钮的尺寸
   * @en Size of the button
   * @defaultValue default
   */
  size?: 'mini' | 'small' | 'default' | 'large';
  /**
   * @zh 按钮形状，`circle` - 圆形， `round` - 全圆角， `square` - 长方形
   * @en Three button shapes are available: `circle`, `round` and `square`
   * @defaultValue square
   */
  shape?: 'circle' | 'round' | 'square';
  /**
   * @zh 添加跳转链接，设置此属性，button表现跟a标签一致
   * @en The button behaves like `<a>` with href as target url.
   */
  href?: string;
  /**
   * @zh a 链接的 target 属性，href 存在时生效
   * @en The target attribute of the link, which takes effect when href exists.
   */
  target?: string;
  /**
   * @zh a 链接的原生属性，href 存在时生效
   * @en The native attribute of the link, which takes effect when href exists
   */
  anchorProps?: HTMLProps<HTMLAnchorElement>;
  /**
   * @zh 是否禁用
   * @en Whether to disable the button
   */
  disabled?: boolean;
  /**
   * @zh 按钮是否是加载状态
   * @en Whether the button is in the loading state
   */
  loading?: boolean;
  /**
   * @zh 当 loading 的时候，不改变按钮的宽度。
   * @en The width of the button remains unchanged on loading
   */
  loadingFixedWidth?: boolean;
  /**
   * @zh 设置按钮的图标
   * @en Icon of the button
   */
  icon?: ReactNode;
  /**
   * @zh 只有图标，按钮宽高相等。如果指定 `icon` 且没有 children，`iconOnly` 默认为 true
   * @en Whether to show icon only, in which case the button width and height are equal. If `icon` is specified and there are no children, `iconOnly` defaults to `true`
   */
  iconOnly?: boolean;
  /**
   * @zh 按钮宽度随容器自适应。
   * @en Whether the width of the button should adapt to the container.
   */
  long?: boolean;
  /**
   * @zh 点击按钮的回调
   * @en Callback fired when the button is clicked
   */
  onClick?: (e: Event) => void;
}

export type AnchorButtonProps = {
  href: string;
  target?: string;
  anchorProps?: HTMLProps<HTMLAnchorElement>;
} & BaseButtonProps &
  Omit<React.AnchorHTMLAttributes<any>, 'type' | 'onClick' | 'className'>;

export type FinalButtonProps = {
  /**
   * @zh 按钮原生的 html type 类型
   * @en html button type
   * @defaultValue button
   */
  htmlType?: 'button' | 'submit' | 'reset';
} & BaseButtonProps &
  Omit<React.ButtonHTMLAttributes<any>, 'type' | 'onClick' | 'className'>;

/**
 * @title Button
 */
export type ButtonProps = Partial<FinalButtonProps & AnchorButtonProps>;

export interface ButtonGroupProps {
  style?: CSSProperties;
  className?: string | string[];
  children?: ReactNode;
}
