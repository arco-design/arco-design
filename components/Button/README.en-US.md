`````
Component / General

# Button

A button is a command component to trigger an operation.
`````

%%Content%%

## API

### Button

|Property|Description|Type|DefaultValue|
|---|---|---|---|
|htmlType|html button type|`'button' \| 'submit' \| 'reset'`|`button`|
|style|Additional style|`CSSProperties`|`-`|
|className|Additional css class|`string \| string[]`|`-`|
|type|A variety of button types are available: `primary`, `secondary`, `dashed`,`text`, `linear` and `default` which is the secondary.|`'default' \| 'primary' \| 'secondary' \| 'dashed' \| 'text' \| 'outline'`|`default`|
|status|Status of the button|`'warning' \| 'danger' \| 'success' \| 'default'`|`default`|
|size|Size of the button|`'mini' \| 'small' \| 'default' \| 'large'`|`default`|
|shape|Three button shapes are available: `circle`, `round` and `square`|`'circle' \| 'round' \| 'square'`|`square`|
|href|The button behaves like `<a>` with href as target url.|`string`|`-`|
|target|The target attribute of the link, which takes effect when href exists.|`string`|`-`|
|anchorProps|The native attribute of the link, which takes effect when href exists|`HTMLProps<HTMLAnchorElement>`|`-`|
|disabled|Whether to disable the button|`boolean`|`-`|
|loading|Whether the button is in the loading state|`boolean`|`-`|
|loadingFixedWidth|The width of the button remains unchanged on loading|`boolean`|`-`|
|icon|Icon of the button|`ReactNode`|`-`|
|iconOnly|Whether to show icon only, in which case the button width and height are equal. If `icon` is specified and there are no children, `iconOnly` defaults to `true`|`boolean`|`-`|
|long|Whether the width of the button should adapt to the container.|`boolean`|`-`|
|onClick|Callback fired when the button is clicked|`(e: Event) => void`|`-`|
