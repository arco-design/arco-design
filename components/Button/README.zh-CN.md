`````
组件 / 通用

# 按钮 Button

按钮是一种命令组件，可发起一个即时操作。
`````

%%Content%%

## API

### Button

|参数名|描述|类型|默认值|
|---|---|---|---|
|htmlType|按钮原生的 html type 类型|`'button' \| 'submit' \| 'reset'`|`button`|
|style|节点样式|`CSSProperties`|`-`|
|className|节点类名|`string \| string[]`|`-`|
|type|按钮主要分为六种按钮类型：主要按钮、次级按钮、虚框按钮、文字按钮、线性按钮，`default` 为次级按钮。|`'default' \| 'primary' \| 'secondary' \| 'dashed' \| 'text' \| 'outline'`|`default`|
|status|按钮状态|`'warning' \| 'danger' \| 'success' \| 'default'`|`default`|
|size|按钮的尺寸|`'mini' \| 'small' \| 'default' \| 'large'`|`default`|
|shape|按钮形状，`circle` - 圆形， `round` - 全圆角， `square` - 长方形|`'circle' \| 'round' \| 'square'`|`square`|
|href|添加跳转链接，设置此属性，button表现跟a标签一致|`string`|`-`|
|target|a 链接的 target 属性，href 存在时生效|`string`|`-`|
|anchorProps|a 链接的原生属性，href 存在时生效|`HTMLProps<HTMLAnchorElement>`|`-`|
|disabled|是否禁用|`boolean`|`-`|
|loading|按钮是否是加载状态|`boolean`|`-`|
|loadingFixedWidth|当 loading 的时候，不改变按钮的宽度。|`boolean`|`-`|
|icon|设置按钮的图标|`ReactNode`|`-`|
|iconOnly|只有图标，按钮宽高相等。如果指定 `icon` 且没有 children，`iconOnly` 默认为 true|`boolean`|`-`|
|long|按钮宽度随容器自适应。|`boolean`|`-`|
|onClick|点击按钮的回调|`(e: Event) => void`|`-`|
