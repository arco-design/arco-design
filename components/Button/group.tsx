import React, { useContext } from 'react';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { ButtonGroupProps } from './interface';

function Group(props: ButtonGroupProps, ref) {
  const { className, style, children, ...rest } = props;
  const { getPrefixCls } = useContext(ConfigContext);
  const prefixCls = getPrefixCls('btn-group');
  const classNames = cs(prefixCls, className);

  return (
    <div ref={ref} className={classNames} style={style} {...rest}>
      {children}
    </div>
  );
}

const GroupComponent = React.forwardRef<unknown, ButtonGroupProps>(Group);

GroupComponent.displayName = 'ButtonGroup';

export default GroupComponent;

export { ButtonGroupProps };
