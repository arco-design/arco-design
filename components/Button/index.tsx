import React, { useState, useContext, useEffect, createRef, ReactNode, forwardRef } from 'react';
import IconLoading from '../../icon/react-icon/IconLoading';
import Group from './group';
import cs from '../_util/classNames';
import { ConfigContext } from '../ConfigProvider';
import { ButtonProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const regexTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;

function processChildren(children?: ReactNode) {
  const childrenList = [];
  let isPrevChildPure = false;
  React.Children.forEach(children, (child) => {
    const isCurrentChildPure = typeof child === 'string' || typeof child === 'number';
    if (isCurrentChildPure && isPrevChildPure) {
      const lastIndex = childrenList.length - 1;
      const lastChild = childrenList[lastIndex];
      childrenList[lastIndex] = `${lastChild}${child}`;
    } else {
      childrenList.push(child);
    }
    isPrevChildPure = isCurrentChildPure;
  });
  return React.Children.map(childrenList, (child) =>
    typeof child === 'string' ? <span>{child}</span> : child
  );
}

const defaultProps: ButtonProps = {
  htmlType: 'button',
  type: 'default',
  shape: 'square',
};

function Button(baseProps: ButtonProps, ref) {
  const {
    getPrefixCls,
    size: ctxSize,
    autoInsertSpaceInButton,
    componentConfig,
  } = useContext(ConfigContext);
  const props = useMergeProps<ButtonProps>(baseProps, defaultProps, componentConfig?.Button);
  const {
    style,
    className,
    children,
    htmlType,
    type,
    status,
    size,
    shape,
    href,
    anchorProps,
    disabled,
    loading,
    loadingFixedWidth,
    icon,
    iconOnly,
    onClick,
    long,
    ...rest
  } = props;

  const iconNode = loading ? <IconLoading /> : icon;

  const [isTwoCNChar, setIsTwoCNChar] = useState(false);
  const buttonRef = ref || createRef();

  useEffect(() => {
    if (autoInsertSpaceInButton && buttonRef && buttonRef.current) {
      const textContent = buttonRef.current.textContent;
      if (regexTwoCNChar.test(textContent)) {
        if (!isTwoCNChar) {
          setIsTwoCNChar(true);
        }
      } else if (isTwoCNChar) {
        setIsTwoCNChar(false);
      }
    }
  }, [buttonRef, autoInsertSpaceInButton]);

  const prefixCls = getPrefixCls('btn');
  const _type = type === 'default' ? 'secondary' : type;
  const classNames = cs(
    prefixCls,
    `${prefixCls}-${_type}`,
    `${prefixCls}-size-${size || ctxSize}`,
    `${prefixCls}-shape-${shape}`,
    {
      [`${prefixCls}-long`]: long,
      [`${prefixCls}-status-${status}`]: status,
      [`${prefixCls}-loading-fixed-width`]: loadingFixedWidth,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-link`]: href,
      [`${prefixCls}-icon-only`]: iconOnly || (!children && children !== 0 && iconNode),
      [`${prefixCls}-disabled`]: disabled,
      [`${prefixCls}-two-chinese-chars`]: isTwoCNChar,
    },
    className
  );

  const handleClick: React.MouseEventHandler<HTMLElement> = (event: any): void => {
    if (loading) {
      typeof event?.preventDefault === 'function' && event.preventDefault();
      return;
    }
    onClick && onClick(event);
  };

  const InnerContent = (
    <>
      {iconNode}
      {processChildren(children)}
    </>
  );

  if (href) {
    const _anchorProps = { ...anchorProps };
    if (disabled) {
      delete _anchorProps.href;
    } else {
      _anchorProps.href = href;
    }
    return (
      <a
        ref={buttonRef}
        {...rest}
        {..._anchorProps}
        style={style}
        className={classNames}
        onClick={handleClick}
      >
        {InnerContent}
      </a>
    );
  }

  return (
    <button
      ref={buttonRef}
      {...rest}
      style={style}
      className={classNames}
      type={htmlType}
      disabled={disabled}
      onClick={handleClick}
    >
      {InnerContent}
    </button>
  );
}

const ForwardRefButton = forwardRef<unknown, ButtonProps>(Button);

const ButtonComponent = ForwardRefButton as typeof ForwardRefButton & {
  __BYTE_BUTTON: boolean;
  Group: typeof Group;
};

ButtonComponent.__BYTE_BUTTON = true;

ButtonComponent.Group = Group;

ButtonComponent.displayName = 'Button';

export default ButtonComponent;

export { ButtonProps };
