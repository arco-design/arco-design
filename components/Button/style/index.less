@import './token.less';

@btn-prefix-cls: ~'@{prefix}-btn';

.btn-type(@type) {
  .@{btn-prefix-cls}-@{type}:not(.@{btn-prefix-cls}-disabled) {
    background-color: ~'@{btn-@{type}-color-bg}';
    color: ~'@{btn-@{type}-color-text}';
    border: @btn-size-default-border-width ~'@{btn-@{type}-border-style}' ~'@{btn-@{type}-color-border}';

    &:not(.@{btn-prefix-cls}-loading):hover {
      border-color: ~'@{btn-@{type}-color-border_hover}';
      color: ~'@{btn-@{type}-color-text_hover}';
      background-color: ~'@{btn-@{type}-color-bg_hover}';
    }

    &:not(.@{btn-prefix-cls}-loading):active {
      border-color: ~'@{btn-@{type}-color-border_active}';
      color: ~'@{btn-@{type}-color-text_active}';
      background-color: ~'@{btn-@{type}-color-bg_active}';
    }

    &:focus-visible {
      box-shadow: 0 0 0 @btn-box-shadow-radius ~'@{btn-@{type}-color-box-shadow}';
    }
  }

  .@{btn-prefix-cls}-@{type}.@{btn-prefix-cls}-disabled {
    background-color: ~'@{btn-@{type}-color-bg_disabled}';
    color: ~'@{btn-@{type}-color-text_disabled}';
    border: @btn-size-default-border-width ~'@{btn-@{type}-border-style}' ~'@{btn-@{type}-color-border_disabled}';
    cursor: not-allowed;
  }
}

.btn-status(@type, @status) {
  .@{btn-prefix-cls}-@{type}.@{btn-prefix-cls}-status-@{status}:not(.@{btn-prefix-cls}-disabled) {
    background-color: ~'@{btn-@{type}-color-bg_@{status}}';
    color: ~'@{btn-@{type}-color-text_@{status}}';
    border-color: ~'@{btn-@{type}-color-border_@{status}}';

    &:not(.@{btn-prefix-cls}-loading):hover {
      border-color: ~'@{btn-@{type}-color-border_@{status}_hover}';
      color: ~'@{btn-@{type}-color-text_@{status}_hover}';
      background-color: ~'@{btn-@{type}-color-bg_@{status}_hover}';
    }

    &:not(.@{btn-prefix-cls}-loading):active {
      border-color: ~'@{btn-@{type}-color-border_@{status}_active}';
      color: ~'@{btn-@{type}-color-text_@{status}_active}';
      background-color: ~'@{btn-@{type}-color-bg_@{status}_active}';
    }

    &:focus-visible {
      box-shadow: 0 0 0 @btn-box-shadow-radius ~'@{btn-color-box-shadow_@{status}}';
    }
  }

  .@{btn-prefix-cls}-@{type}.@{btn-prefix-cls}-status-@{status}.@{btn-prefix-cls}-disabled {
    color: ~'@{btn-@{type}-color-text_@{status}_disabled}';
    background-color: ~'@{btn-@{type}-color-bg_@{status}_disabled}';
    border: @btn-size-default-border-width ~'@{btn-@{type}-border-style}' ~'@{btn-@{type}-color-border_@{status}_disabled}';
  }
}

.btn-size(@size) {
  .@{btn-prefix-cls}-size-@{size} {
    padding: 0 ~'@{btn-size-@{size}-padding-horizontal}';
    font-size: ~'@{btn-size-@{size}-font-size}';
    height: ~'@{btn-size-@{size}-height}';
    border-radius: ~'@{btn-size-@{size}-radius}';
    // line-height: calc(~'@{size-@{size}}' - ~'@{btn-size-@{size}-border-width}' * 2);

    > svg + span,
    > span + svg {
      margin-left: ~'@{btn-size-@{size}-icon-spacing}';
    }

    svg {
      vertical-align: ~'@{btn-size-@{size}-icon-vertical-align}';
    }

    &.@{btn-prefix-cls}-rtl {
      > svg + span,
      > span + svg {
        margin-left: 0;
        margin-right: ~'@{btn-size-@{size}-icon-spacing}';
      }
    }
  }

  @bsp: ~'btn-size-@{size}-padding-horizontal';
  @bsf: ~'btn-size-@{size}-font-size';
  @bss: ~'btn-size-@{size}-icon-spacing';

  .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-loading-fixed-width.@{btn-prefix-cls}-loading {
    padding-left: @@bsp - ((@@bsf + @@bss) / 2);
    padding-right: @@bsp - ((@@bsf + @@bss) / 2);
  }

  .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-icon-only {
    width: ~'@{size-@{size}}';
    height: ~'@{size-@{size}}';
    padding: 0;
  }

  .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-circle {
    width: ~'@{size-@{size}}';
    height: ~'@{size-@{size}}';
    padding: 0;
    text-align: center;
    border-radius: @radius-circle;
  }

  .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-round {
    border-radius: calc(~'@{size-@{size}}' * 0.5);
  }

  .@{btn-prefix-cls}-group {
    .@{btn-prefix-cls}-size-@{size}:first-child {
      border-radius: ~'@{btn-size-@{size}-radius}' 0 0 ~'@{btn-size-@{size}-radius}';
    }

    .@{btn-prefix-cls}-size-@{size}:last-child {
      border-radius: 0 ~'@{btn-size-@{size}-radius}' ~'@{btn-size-@{size}-radius}' 0;
    }

    .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-round:first-child {
      border-radius: calc(~'@{size-@{size}}' * 0.5) 0 0 calc(~'@{size-@{size}}' * 0.5);
    }

    .@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-round:last-child {
      border-radius: 0 calc(~'@{size-@{size}}' * 0.5) calc(~'@{size-@{size}}' * 0.5) 0;
    }

    .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-size-@{size}:first-child {
      border-radius: 0 ~'@{btn-size-@{size}-radius}' ~'@{btn-size-@{size}-radius}' 0;
    }

    .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-size-@{size}:last-child {
      border-radius: ~'@{btn-size-@{size}-radius}' 0 0 ~'@{btn-size-@{size}-radius}';
    }

    .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-round:first-child {
      border-radius: 0 calc(~'@{size-@{size}}' * 0.5) calc(~'@{size-@{size}}' * 0.5) 0;
    }

    .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-size-@{size}.@{btn-prefix-cls}-shape-round:last-child {
      border-radius: calc(~'@{size-@{size}}' * 0.5) 0 0 calc(~'@{size-@{size}}' * 0.5);
    }
  }
}

.@{btn-prefix-cls} {
  display: inline-block;
  position: relative;
  outline: none;
  font-weight: @btn-font-weight;
  appearance: none;
  user-select: none;
  cursor: pointer;
  white-space: nowrap;
  transition: all @transition-duration-1 @transition-timing-function-linear;
  box-sizing: border-box;
  line-height: @line-height-base;

  > a:only-child {
    color: currentColor;
  }

  &:active {
    transition: none;
  }

  &:empty {
    display: inline-block;
    vertical-align: bottom;
  }

  &-long {
    display: block;
    width: 100%;
  }

  // link
  &-link {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-decoration: none;

    &:not([href]) {
      color: var(~'@{arco-cssvars-prefix}-color-text-4');
    }

    &:hover {
      text-decoration: none;
    }
  }

  &-loading {
    cursor: default;
    position: relative;

    &::before {
      content: '';
      position: absolute;
      top: -1px;
      right: -1px;
      bottom: -1px;
      left: -1px;
      z-index: 1;
      display: block;
      background: #fff;
      border-radius: inherit;
      opacity: 0.4;
      transition: opacity @transition-duration-1 @transition-timing-function-linear;
      pointer-events: none;
    }
  }

  &-loading-fixed-width {
    transition: none;
  }

  &-two-chinese-chars > *:not(svg) {
    letter-spacing: 0.3em;
    margin-right: -0.3em;
  }

  a&-icon-only {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    vertical-align: top;
  }
}

.btn-type(outline);
.btn-status(outline, warning);
.btn-status(outline, danger);
.btn-status(outline, success);

.btn-type(primary);
.btn-status(primary, warning);
.btn-status(primary, danger);
.btn-status(primary, success);

.btn-type(secondary);
.btn-status(secondary, warning);
.btn-status(secondary, danger);
.btn-status(secondary, success);

.btn-type(dashed);
.btn-status(dashed, warning);
.btn-status(dashed, danger);
.btn-status(dashed, success);

.btn-type(text);
.btn-status(text, warning);
.btn-status(text, danger);
.btn-status(text, success);

.btn-size(mini);
.btn-size(small);
.btn-size(default);
.btn-size(large);

.@{btn-prefix-cls}-group {
  display: inline-block;

  .@{btn-prefix-cls}-outline:not(:first-child),
  .@{btn-prefix-cls}-dashed:not(:first-child) {
    margin-left: -1px;
  }

  .@{btn-prefix-cls}-primary:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_hover;
  }

  .@{btn-prefix-cls}-secondary:not(:last-child) {
    border-right: 1px solid @btn-secondary-color-bg_hover;
  }

  .@{btn-prefix-cls}-text:not(:last-child) {
    border-right: 1px solid @btn-text-color-border_hover;
  }

  .@{btn-prefix-cls}-status-warning:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_warning_hover;
    &.@{btn-prefix-cls}-text {
      border-right: 1px solid @btn-text-color-bg_warning_hover;
    }
  }

  .@{btn-prefix-cls}-status-danger:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_danger_hover;
    &.@{btn-prefix-cls}-text {
      border-right: 1px solid @btn-text-color-bg_danger_hover;
    }
  }

  .@{btn-prefix-cls}-status-success:not(:last-child) {
    border-right: 1px solid @btn-primary-color-bg_success_hover;
    &.@{btn-prefix-cls}-text {
      border-right: 1px solid @btn-text-color-bg_success_hover;
    }
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-outline:not(:first-child),
  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-dashed:not(:first-child) {
    margin-left: 0;
    margin-right: -1px;
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-primary:not(:last-child) {
    border-left: 1px solid @btn-primary-color-bg_hover;
    border-right: 0;
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-secondary:not(:last-child) {
    border-left: 1px solid @btn-secondary-color-bg_hover;
    border-right: 0;
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-text:not(:last-child) {
    border-left: 1px solid @btn-text-color-border_hover;
    border-right: 0;
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-status-warning:not(:last-child) {
    border-left: 1px solid @btn-primary-color-bg_warning_hover;
    border-right: 0;
    &.@{btn-prefix-cls}-text {
      border-left: 1px solid @btn-text-color-bg_warning_hover;
    }
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-status-danger:not(:last-child) {
    border-left: 1px solid @btn-primary-color-bg_danger_hover;
    border-right: 0;
    &.@{btn-prefix-cls}-text {
      border-left: 1px solid @btn-text-color-bg_danger_hover;
    }
  }

  .@{btn-prefix-cls}-rtl.@{btn-prefix-cls}-status-success:not(:last-child) {
    border-left: 1px solid @btn-primary-color-bg_success_hover;
    border-right: 0;
    &.@{btn-prefix-cls}-text {
      border-left: 1px solid @btn-text-color-bg_success_hover;
    }
  }

  .@{btn-prefix-cls}-outline,
  .@{btn-prefix-cls}-dashed {
    &:hover,
    &:active {
      z-index: 2;
    }
  }

  .@{btn-prefix-cls}:not(:first-child):not(:last-child) {
    border-radius: 0;
  }
}

.@{btn-prefix-cls}-rtl {
  direction: rtl;
}

@{arco-theme-tag} {
  &[arco-theme='dark'] {
    .@{btn-prefix-cls}-primary.@{btn-prefix-cls}-disabled {
      color: rgba(255, 255, 255, 30%);
    }
  }
}
