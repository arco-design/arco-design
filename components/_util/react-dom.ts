import { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { isObject } from './is';

type CreateRootFnType = (container: Element | DocumentFragment) => {
  render: (container: ReactElement) => void;
  unmount: () => void;
  _unmount: () => void;
};

const __SECRET_INTERNALS__ = '__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED';

const CopyReactDOM = ReactDOM as typeof ReactDOM & {
  createRoot: CreateRootFnType;
  // https://github.com/facebook/react/blob/4ff5f5719b348d9d8db14aaa49a48532defb4ab7/packages/react-dom/src/client/ReactDOM.js#L181
  [__SECRET_INTERNALS__]: {
    usingClientEntryPoint?: boolean;
  };
};

let copyRender: (
  app: ReactElement,
  container: Element | DocumentFragment
) => {
  render: (container: ReactElement) => void;
  _unmount: () => void;
};

const isReact18 = Number(CopyReactDOM.version?.split('.')[0]) > 17;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
  // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
  // Avoid console warning
  if (isObject(CopyReactDOM[__SECRET_INTERNALS__])) {
    CopyReactDOM[__SECRET_INTERNALS__].usingClientEntryPoint = skipWarning;
  }
};

let createRoot: CreateRootFnType;
try {
  createRoot = CopyReactDOM.createRoot;
} catch (_) {
  //
}

if (isReact18 && createRoot) {
  copyRender = (app: ReactElement, container: Element | DocumentFragment) => {
    updateUsingClientEntryPoint(true);
    const root = createRoot(container);
    updateUsingClientEntryPoint(false);

    root.render(app);

    root._unmount = function () {
      setTimeout(() => {
        root?.unmount?.();
      });
    };
    return root;
  };
} else {
  copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
    CopyReactDOM.render(app, container);

    return {
      render: (app: ReactElement) => {
        CopyReactDOM.render(app, container);
      },
      _unmount() {
        CopyReactDOM.unmountComponentAtNode(container);
      },
    };
  };
}

export const render = copyRender;
