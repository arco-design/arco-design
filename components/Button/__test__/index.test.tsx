import React from 'react';
import { act } from 'react-dom/test-utils';
import { render, fireEvent } from '../../../tests/util';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Button from '..';

mountTest(Button);
componentConfigTest(Button, 'Button');
componentConfigTest(Button.Group, 'Button.Group');

describe('button', () => {
  it('click callback correctly', () => {
    const mockFn = jest.fn();
    const component = render(<Button onClick={mockFn} />);
    const button = component.querySelector('button') as HTMLElement;
    fireEvent.click(button);
    const mockFnCallLength = mockFn.mock.calls.length;
    expect(mockFnCallLength).toBe(1);

    component.rerender(<Button onClick={mockFn} disabled />);

    fireEvent.click(button);
    expect(mockFn.mock.calls.length).toBe(mockFnCallLength);
  });

  it('render multiple children correctly', () => {
    const { container } = render(
      <Button>
        1{'  '}2{'  '}3{'  '}
      </Button>
    );
    expect(container.firstElementChild?.textContent).toEqual('1  2  3  ');
  });

  it('use context autoInsertSpaceInButton correctly', () => {
    const mockText = '测试';
    const realContext = React.useContext;
    const mockContext = jest.fn().mockReturnValue({
      autoInsertSpaceInButton: true,
      getPrefixCls: (componentName) => `test-${componentName}`,
    });

    const prefixCls = `.test-btn`;
    React.useContext = mockContext;

    const button = render(<Button>{mockText}</Button>);
    expect(button.querySelector(prefixCls)).toHaveClass('test-btn-two-chinese-chars');

    React.useContext = realContext;
  });

  it('render href type correctly', () => {
    const button = render(
      <Button type="primary" href="https://arco.design">
        测试
      </Button>
    );
    expect(button.find('button')).toHaveLength(0);
    expect(button.find('a')).toHaveLength(1);
    expect(button.querySelector('a')?.getAttribute('href')).not.toBeUndefined();

    act(() => {
      button.rerender(<Button type="primary" href="https://arco.design" disabled />);
    });

    expect(button.querySelector('a')?.getAttribute('href')).toBeNull();
  });
});
