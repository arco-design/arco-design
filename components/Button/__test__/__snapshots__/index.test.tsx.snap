// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Button default 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button set className globally 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square global-default-class"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button set className globally and component set className itself. 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square component-class"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button set data-* property 1`] = `
<button
  class="arco-btn arco-btn-secondary arco-btn-size-default arco-btn-shape-square"
  data-test-name="Button"
  type="button"
/>
`;

exports[`ConfigProvider componentConfig.Button.Group default 1`] = `
<div
  class="arco-btn-group"
/>
`;

exports[`ConfigProvider componentConfig.Button.Group set className globally 1`] = `
<div
  class="arco-btn-group"
/>
`;

exports[`ConfigProvider componentConfig.Button.Group set className globally and component set className itself. 1`] = `
<div
  class="arco-btn-group component-class"
/>
`;

exports[`ConfigProvider componentConfig.Button.Group set data-* property 1`] = `
<div
  class="arco-btn-group"
  data-test-name="Button.Group"
/>
`;
